libgalois/include/galois/ParallelSTL.h

```diff
@@ -305,6 +306,61 @@ std::enable_if_t<!std::is_scalar<internal::Val_ty<I>>::value> destroy(I first,
 template <class I>
 std::enable_if_t<std::is_scalar<internal::Val_ty<I>>::value> destroy(I, I) {}
 
+/**
+ * Does a partial sum from first -> last and writes the results to the d_first
+ * iterator.
+ */
+template <class InputIt, class OutputIt>
+OutputIt partial_sum(InputIt first, InputIt last, OutputIt d_first) {
+  using ValueType = typename std::iterator_traits<InputIt>::value_type;
+
+  size_t sizeOfVector = std::distance(first, last);
+
+  // optimization possible here: if num blocks = 1, then no need for 2 passes
+  const size_t numBlocks = galois::getActiveThreads();
+  const size_t blockSize = (sizeOfVector + numBlocks - 1) / numBlocks;
+  assert(numBlocks * blockSize >= sizeOfVector);
+
+  std::vector<ValueType> localSums(numBlocks);
+
+  // get the block sums
+  galois::do_all(
+      galois::iterate((size_t)0, numBlocks), [&](const size_t& block) {
+        ValueType lsum = 0;
+        // loop over block
+        size_t blockEnd = std::min((block + 1) * blockSize, sizeOfVector);
+        for (size_t i = block * blockSize; i < blockEnd; i++) {
+          lsum += *(first + i);
```

Comment:
It's better to partially accumulate the blocks here. In fact, I would prefer if you can just std::partial_sort() on each block in each thread.

Reply:
ack